library/std/src/sys/pal/unix/locks/queue_rwlock.rs

```diff
@@ -0,0 +1,555 @@
+//! Efficient read-write locking without `pthread_rwlock_t`.
+//!
+//! The readers-writer lock provided by the `pthread` library has a number of
+//! problems which make it a suboptimal choice for `std`:
+//!
+//! * It is non-movable, so it needs to be allocated (lazily, to make the
+//! constructor `const`).
+//! * `pthread` is an external library, meaning the fast path of acquiring an
+//! uncontended lock cannot be inlined.
+//! * Some platforms (at least glibc before version 2.25) have buggy implementations
+//! that can easily lead to undefined behaviour in safe Rust code when not properly
+//! guarded against.
```

Comment:
> at least glibc before version 2.25

Don't all platforms that run glibc use the futex_rwlock implementation anyway?

Reply:
No (I think), cause Hurd is a thing, but regardless, we still do the extra checks in the pthread version, probably to guard against other bad implementations.